### ISSUE TITLE
Avoid large queries on sites with thousands of categories (core WP issue)

### ISSUE THREAD

**Issue by mikejolley** [woocommerce/woocommerce#12773]
On sites with lots of categories (thousands) a few areas in WordPress admin try to grab them all for display which can be inefficient.

We either need to implement caching, remove these areas over a certain threshold, or build in search instead of grabbing all at once.

Affects:

- [ ] Category dropdown on product list page #12471
- [ ] Category meta box on edit product page #12769 
- <del>Category box on appearance > menus screen</del>

Basically, anywhere where 'all categories' are displayed.

![2017-01-06 at 10 51](https://cloud.githubusercontent.com/assets/90977/21715920/2277c396-d3ff-11e6-80d0-53120ff29f9a.png)

Look into ajax or API based search fields instead.

**Comment by spencerfinnell** [comment]
May want to take direction from how search for items is offered/handled in the Menus page: https://cloudup.com/c3VJtchxXnR

The thing you lose with searching is hierarchy (in my screenshot "Child" has no visible indication of its parent).

**Comment by mikejolley** [comment]
The menus page is paginated so we don't need to work on that one any more.

**Comment by WPprodigy** [comment]
Ran into this issue on a large site. 2,034 categories, 634,190 products. You can't load the product's page w/ the filters on it (server has a lot of resources too). 

Commenting out https://github.com/woocommerce/woocommerce/blob/a9058fe10bca25e5bd7330f8878a2441ab8d4606/includes/admin/class-wc-admin-post-types.php#L1459 helps the page load, and under 2 seconds.

Here is a screenshot from another site where the products admin would load w/ the filters: http://cld.wthms.co/E0f7Jm. This is only w/ a partial number of the products (41k) and categories that they would like to have.

____

This improvement would help a lot for large stores. In the short term, maybe we could add a filter to product_filters() so the the category filtering can be prevented without hacking core?

Long term, we really only need to implement a solution for this for stores that have over a certain amount of categories. For the majority of users, this will never affect them. After a threshold, I'd vote on the search option. Losing hierarchy is worth the performance, but in theory it should be possible to keep hierarchy using parent ids.

**Comment by mikejolley** [comment]
@WPprodigy Yeah I don't think we should change that then, not until we have our own edit page UI to replace the meta boxes.

The list view we can fix/ajaxify. PR in is.
